File: src/LTAdmin/Services/Business/MentionHelper.cs
using LTAdmin.Models.Entities;

namespace LTAdmin.Services.Business;

/// <summary>
/// Helpers partagés par les bulletins et les résultats finaux :
/// lecture de la grille des mentions, appréciations par défaut et rangs.
/// </summary>
public static class MentionHelper
{
    /// <summary>
    /// Trouve la mention pour une moyenne selon la grille : intervalles [INF, SUP[,
    /// sauf la borne supérieure maximale (20) qui est incluse. La grille doit être
    /// triée par INF décroissant (ordre du dépôt).
    /// </summary>
    public static GrilleMention? Trouver(List<GrilleMention> grille, double? moyenne)
    {
        if (!moyenne.HasValue) return null;
        var value = moyenne.Value;
        GrilleMention? fallback = null;
        foreach (var ligne in grille)
        {
            var inf = ligne.Inf ?? 0d;
            var sup = ligne.Sup ?? 20d;
            fallback ??= ligne;
            if (value >= inf && (value < sup || sup >= 20d))
                return ligne;
        }
        return fallback;
    }

    /// <summary>Appréciation par défaut selon la mention (modifiable ensuite à la main).</summary>
    public static string AppreciationPour(string? mention, bool? admis)
    {
        if (string.IsNullOrWhiteSpace(mention))
            return admis == false ? "Résultats insuffisants. Doit fournir davantage d’efforts." : "Résultats en cours d’évaluation.";
        var text = mention.Trim().ToLowerInvariant();
        if (text.Contains("très bien") || text.Contains("tres bien")) return "Excellent travail. Félicitations !";
        if (text.Contains("bien")) return "Bon travail. Continuez ainsi.";
        if (text.Contains("assez bien")) return "Travail assez bien. Peut encore progresser.";
        if (text.Contains("passable")) return "Travail passable. Des efforts restent nécessaires.";
        return "Résultats insuffisants. Doit fournir davantage d’efforts.";
    }

    /// <summary>
    /// Classement « competition » (1, 2, 2, 4) sur des valeurs triées par ordre
    /// décroissant. Les valeurs nulles ne sont pas classées (rang 0).
    /// </summary>
    public static Dictionary<int, int> RangsCompetition(List<(int Id, double? Valeur)> valeurs)
    {
        var rangs = new Dictionary<int, int>();
        var classees = valeurs
            .Where(v => v.Valeur.HasValue)
            .OrderByDescending(v => v.Valeur!.Value)
            .ToList();
        var rang = 0;
        var position = 0;
        double? precedent = null;
        foreach (var item in classees)
        {
            position++;
            if (!precedent.HasValue || Math.Abs(item.Valeur!.Value - precedent.Value) > 0.0001)
                rang = position;
            rangs[item.Id] = rang;
            precedent = item.Valeur;
        }
        foreach (var item in valeurs.Where(v => !v.Valeur.HasValue))
            rangs[item.Id] = 0;
        return rangs;
    }
}

File: ltadmin/services/business/mention_helper.py
"""Helpers partagés par les bulletins et les résultats finaux.

Lecture de la grille des mentions, appréciations par défaut et rangs
« competition » (1, 2, 2, 4 ; non classés = 0).
"""

from __future__ import annotations

from typing import Dict, List, Optional, Tuple

from ltadmin.models.entities import GrilleMention


def trouver(grille: List[GrilleMention], moyenne: Optional[float]) -> Optional[GrilleMention]:
    """Trouve la mention pour une moyenne : intervalles [INF, SUP[, sauf la
    borne supérieure maximale (20) qui est incluse. La grille doit être triée
    par INF décroissant (ordre du dépôt)."""
    if moyenne is None:
        return None
    value = moyenne
    fallback: Optional[GrilleMention] = None
    for ligne in grille:
        inf = ligne.inf if ligne.inf is not None else 0.0
        sup = ligne.sup if ligne.sup is not None else 20.0
        if fallback is None:
            fallback = ligne
        if value >= inf and (value < sup or sup >= 20.0):
            return ligne
    return fallback


def appreciation_pour(mention: Optional[str], admis: Optional[bool]) -> str:
    """Appréciation par défaut selon la mention (modifiable ensuite à la main)."""
    if not mention or not mention.strip():
        if admis is False:
            return "Résultats insuffisants. Doit fournir davantage d’efforts."
        return "Résultats en cours d’évaluation."
    text = mention.strip().lower()
    if "très bien" in text or "tres bien" in text:
        return "Excellent travail. Félicitations !"
    if "bien" in text:
        return "Bon travail. Continuez ainsi."
    if "assez bien" in text:
        return "Travail assez bien. Peut encore progresser."
    if "passable" in text:
        return "Travail passable. Des efforts restent nécessaires."
    return "Résultats insuffisants. Doit fournir davantage d’efforts."


def rangs_competition(valeurs: List[Tuple[int, Optional[float]]]) -> Dict[int, int]:
    """Classement « competition » (1, 2, 2, 4) sur des valeurs décroissantes.

    Les valeurs nulles ne sont pas classées (rang 0).
    """
    rangs: Dict[int, int] = {}
    classees = sorted(
        (item for item in valeurs if item[1] is not None),
        key=lambda item: item[1], reverse=True)
    rang = 0
    position = 0
    precedent: Optional[float] = None
    for identifiant, valeur in classees:
        position += 1
        if precedent is None or abs(valeur - precedent) > 0.0001:
            rang = position
        rangs[identifiant] = rang
        precedent = valeur
    for identifiant, valeur in valeurs:
        if valeur is None:
            rangs[identifiant] = 0
    return rangs
